source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+   * @param position supplies the position.
+   * @param n supplies the number of bits to extract.
+   * @param input supplies the
+   */
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)>
+  static IpType extractBits(int p, int n, IpType input) {
+    // The IP's are stored in network byte order. By shifting the value to the left by p bits, the
+    // bits between 0 and p-1 are zero'd out. Then to get the n bits, shift the IP back by the
+    // address_size minus the number of desired bits.
+    return input << p >> (address_size - n);
+  }
+
+  /**
+   * Removes n bits from input starting at 0.
+   * @param n supplies the number of bits to remove.
+   * @param input supplies the
+   */
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)>
+  static IpType removeBits(int n, IpType input) {
+    // The IP's are stored in network byte order. By shifting the value to the left by n bits and
+    // back, this zeros out all bits between 0 and n-1.
+    return input << n >> n;
+  }
+
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)> struct IpPrefixes {
+
+    /**
+     * @return -1 if the current object is less than b. 0 if they are the same. 1
+     * if b is smaller than the current object.
+     */
+    int compare(const IpPrefixes& b) const {
+      {
+        if (ip_ < b.ip_) {
+          return -1;
+        } else if (ip_ > b.ip_) {
+          return 1;
+        } else if (length_ < b.length_) {
+          return -1;
+        } else if (length_ > b.length_) {
+          return 1;
+        } else {
+          return 0;
+        }
+      }
+    }
+
+    bool operator<(const IpPrefixes& b) const { return (this->compare(b) == -1); }
+
+    /**
+     * @return true if other is a prefix of this.
+     */
+    bool isPrefix(const IpPrefixes& other) {
+      return (length_ == 0 || (length_ <= other.length_ &&
+                               extractBits<IpType, address_size>(0, length_, ip_) ==
+                                   extractBits<IpType, address_size>(0, length_, other.ip_)));
+    }
+
+    // The address represented either in uint32_t or uint128.
+    IpType ip_;
+    // Length of the cidr range.
+    int length_;
+    // TODO(ccaraman): Support more than one tag per entry.
+    // Tag for this entry.
+    std::string tag_;
+  };
+
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)> class LcTrieInternal {
+  public:
+    /**
+     * Level Compressed Trie (LC-Trie) implementation for IP(IPv4 or IPv6) prefixes.
+     *
+     * The C++ implementation of the LcTrie found at
+     * https://github.com/beevek/libkrb/blob/master/krb/lc_trie.hpp was used as a reference.
+     * The implemenation above is based off of S. Nilson's code at
+     * http://www.csc.kth.se/~snilsson/software/router/C/
+     *
+     * Note: This trie handles Ipv4 and Ipv6 separately.
+     * Note: The trie can only support up 524288(2^19) nodes. Please see LcNode for further
```

Comment:
Would you be open to making this unlimited? The global routing table currently contains 700,000 prefixes, so it won't fit within the 2^19 limit. I know Envoy itself is unlikely to ever have that many prefixes configured; but if this implementation can support more nodes, it could be useful to people working on other projects someday.

Reply:
I should note, though, that making this unlimited would require using a `uint64_t` instead of `uint32_t` for `LcNode`, so the LC trie would take up more cache space.

Reply:
I'm am open to it. I don't think we need it now. I've added a comment to change the data structure for LcNode to be uint64_t if any consumer requires it.

Reply:
To summarize our offline conversation, the code will stay with uint32_t as the underlying LcNode structure. 

To avoid getting an out of range exception, I'll add a check at the beginning of every buildRecursive call to ensure the size is > the position being accessed in the Lc-Trie that way if the number of trie_ nodes required is greater than 2^20  the code will throw an exception. If this case is it, then the uint64_t conversation should be revisited.

Once there are larger data sets and perf tests, there can be an effort to see if using uint64_t is worth supporting larger set of prefixes. Another benefit is then there can be some experimentation in determining what the buffer can be.